Source/WebCore/platform/graphics/cocoa/H264UtilitiesCocoa.mm

```diff
@@ -0,0 +1,123 @@
+    reader.read(8);
+    // profile_compatibility;
+    reader.read(8);
+    // AVCLevelIndication;
+    reader.read(8);
+    // bit(6) reserved = '111111'b;
+    // unsigned int(2) lengthSizeMinusOne;
+    size_t lengthSize = (*reader.read<uint8_t>() & 0x3) + 1;
+    // bit(3) reserved = '111'b;
+    // unsigned int(5) numOfSequenceParameterSets;
+    size_t numOfSequenceParameterSets = 0x1f & *reader.read<uint8_t>();
+    if (!numOfSequenceParameterSets)
+        return nullptr;
+
+    constexpr size_t kNALTypeSize = 1;
+    constexpr uint8_t kSPSNAL = 7;
+    constexpr uint8_t kPPSNAL = 8;
+
+    Vector<const uint8_t*> paramSetPtrs;
+    paramSetPtrs.reserveInitialCapacity(numOfSequenceParameterSets + 1);
+    Vector<size_t> paramSetSizes;
+    paramSetSizes.reserveInitialCapacity(numOfSequenceParameterSets + 1);
+    size_t offset = 6;
+    for (size_t index = 0; index < numOfSequenceParameterSets; index++) {
+        auto size = reader.read<uint16_t>();
+        if (!size || *size < kNALTypeSize)
+            return nullptr;
+        offset += 2;
+        auto nalType = reader.read<uint8_t>();
+        if (!nalType || (*nalType & 0x1f) != kSPSNAL)
+            return nullptr;
+        const_cast<uint8_t*>(avcc.subspan(offset).data())[0] &= 0x7f;
+        const_cast<uint8_t*>(avcc.subspan(offset).data())[0] |= 0x60;
+        paramSetPtrs.append(avcc.subspan(offset).data());
+        paramSetSizes.append(*size);
+        offset += *size;
+        reader.skip(8 * (*size - kNALTypeSize));
+    }
+
+    // unsigned int(8) numOfPictureParameterSets;
+    auto numOfPictureParameterSets = reader.read<uint8_t>();
+    offset++;
+    if (!numOfPictureParameterSets || !*numOfPictureParameterSets)
+        return nullptr;
+    for (size_t index = 0; index < *numOfPictureParameterSets; index++) {
+        auto size = reader.read<uint16_t>();
+        if (!size || *size < kNALTypeSize)
+            return nullptr;
+        offset += 2;
+        auto nalType = reader.read<uint8_t>();
+        if (!nalType || (*nalType & 0x1f) != kPPSNAL)
+            return nullptr;
+        const_cast<uint8_t*>(avcc.subspan(offset).data())[0] &= 0x7f;
+        const_cast<uint8_t*>(avcc.subspan(offset).data())[0] |= 0x60;
+        paramSetPtrs.append(avcc.subspan(offset).data());
+        paramSetSizes.append(*size);
+        offset += *size;
+        reader.skip(8 * (*size - kNALTypeSize));
+    }
+
+    CMFormatDescriptionRef rawDescription = nullptr;
+    if (PAL::CMVideoFormatDescriptionCreateFromH264ParameterSets(kCFAllocatorDefault, paramSetPtrs.size(), paramSetPtrs.data(), paramSetSizes.data(), lengthSize, &rawDescription))
+        return nullptr;
+    RetainPtr description = adoptCF(rawDescription);
+    auto dimensions = PAL::CMVideoFormatDescriptionGetDimensions(rawDescription);
+    auto presentationDimensions = PAL::CMVideoFormatDescriptionGetPresentationDimensions(rawDescription, true, true);
+
+    Ref info = VideoInfo::create();
```

Comment:
It seems we should introduce VideoInfoData, do we need a ref counted here?

Reply:
VideoInfo is refcounted and attached to all SamplesBlock